processing/src/main/java/org/apache/druid/query/operator/GlueingPartitioningOperator.java

```diff
@@ -0,0 +1,279 @@
+   * It handles the boundaries of partitions within a single RAC and potentially glues
+   * the first partition of the current RAC with the previous RAC if needed.
+   */
+  private class GluedRACsIterator implements Iterator<RowsAndColumns>
+  {
+    private final RowsAndColumns rac;
+    private final int[] boundaries;
+    private int currentIndex = 0;
+    private boolean firstPartitionHandled = false;
+
+    public GluedRACsIterator(RowsAndColumns rac)
+    {
+      this.rac = rac;
+      ClusteredGroupPartitioner groupPartitioner = rac.as(ClusteredGroupPartitioner.class);
+      if (groupPartitioner == null) {
+        groupPartitioner = new DefaultClusteredGroupPartitioner(rac);
+      }
+      this.boundaries = groupPartitioner.computeBoundaries(partitionColumns);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      return currentIndex < boundaries.length - 1;
+    }
+
+    /**
+     * Retrieves the next partition in the RowsAndColumns. If the first partition has not been handled yet,
+     * it may be glued with the previous RowsAndColumns if the partition columns match.
+     *
+     * @return The next RowsAndColumns partition, potentially glued with the previous one.
+     * @throws NoSuchElementException if there are no more partitions.
+     */
+    @Override
+    public RowsAndColumns next()
+    {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+
+      if (!firstPartitionHandled) {
+        firstPartitionHandled = true;
+        int start = boundaries[currentIndex];
+        int end = boundaries[currentIndex + 1];
+        LimitedRowsAndColumns limitedRAC = new LimitedRowsAndColumns(rac, start, end);
+
+        if (isGlueingNeeded(previousRac, limitedRAC)) {
+          RowsAndColumns gluedRAC = getConcatRacForFirstPartition(previousRac, limitedRAC);
+          ensureMaxRowsMaterializedConstraint(gluedRAC.numRows());
+          previousRac = null;
+          currentIndex++;
+          return gluedRAC;
+        } else {
+          if (previousRac != null) {
+            RowsAndColumns temp = previousRac;
+            previousRac = null;
+            return temp;
+          }
+        }
+      }
+
+      int start = boundaries[currentIndex];
+      int end = boundaries[currentIndex + 1];
+      currentIndex++;
+      return new LimitedRowsAndColumns(rac, start, end);
+    }
+
+    /**
+     * Determines whether gluing is needed between the previous RowsAndColumns and the first partition
+     * of the current RowsAndColumns based on the partition columns. If the columns match, the two RACs
+     * will be glued together.
+     *
+     * @param previousRac The previous RowsAndColumns.
+     * @param firstPartitionOfCurrentRac The first partition of the current RowsAndColumns.
+     * @return true if gluing is needed, false otherwise.
+     */
+    private boolean isGlueingNeeded(RowsAndColumns previousRac, RowsAndColumns firstPartitionOfCurrentRac)
```

Comment:
I wonder if this worths the extra hassle...why not just assume `true` in all cases?
the RACs should be pretty big already; so the extra concat will be amortized away

Reply:
Partitioning will be incorrect if we glue blindly, no?

Whether we glue or not --- that is how the partitioning happens for previous RAC and first boundary of current RAC.

If we don't glue it when needed, it'll not send partitioned data to the receiver.

Reply:
> iirc that had some trouble by retaining references to previous racs....

Yeah